animations: Finish support for fractional iteration counts

mrobinson opened the pull request:
This change also improves support for creating animations with negative
delays, as that is necessary to test support for fractional iteration
lengths.

Fixes: #14858

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #14858 
- [x] There are tests for these changes


<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
animations: Finish support for fractional iteration counts

This change also improves support for creating animations with negative
delays, as that is necessary to test support for fractional iteration
lengths.

This change also adjusts existing Servo animation tests which assumed
that advancing to the exact moment of the end of the animation would be
considered "before the end." With this change, this moment is "after the
end."

Fixes: #14858

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/style/animation.rs

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/24283.

emilio commented on components/style/animation.rs:
```diff
@@ -509,9 +537,10 @@ impl Animation {
-            AnimationState::Pending | AnimationState::Canceled | AnimationState::Paused(_) => {
-                return false
+            AnimationState::Paused(progress) => {
+                return self.on_last_iteration() && progress > self.current_iteration_length();
```
Should it be `>=`?

mrobinson replied:
I've corrected this, which I think was a weird side-effect of the fact that we used to iterate past the last expected iteration when animations finished. I've corrected that and updated to code to reflect it. I think things should be less confusing now. This has also led me to make some improvements to how we calculate `elapsedTime` for animation events. It should be correct now for fractional iteration counts.

emilio commented on components/style/animation.rs:
```diff
@@ -494,12 +502,32 @@ impl Animation {
+    /// Assuming this animation is running, whether or not it is on the last iteration.
+    fn on_last_iteration(&self) -> bool {
+        match self.iteration_state {
+            KeyframesIterationState::Finite(current, max) if current > (max - 1.) => true,
```
Probably just:

```rust
match self.iteration_state {
    Finite(current, max) => current > max - 1.,
    Infinite(..) => false,
}
```

mrobinson replied:
I've corrected this.

Manishearth commented:
r? @emilio

emilio reviewed (approved):
This looks pretty sensible to me.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#24283.

mrobinson commented:
@emilio Thanks for the review! I've pushed a new version of the branch that adds a few more extra changes to improve the way we handle iteration counts in animations and calculate `elapsedTime`. If these look good to you, I can land this change.

emilio commented on components/style/animation.rs:
```diff
@@ -494,36 +502,48 @@ impl Animation {
+    /// The length of the current iteration, a number between > 0 and <= 1, where
+    /// 1 is a full iteration and a number less than that represents an iteration that
+    /// is a fraction of the animation duration.
+    pub fn current_iteration_length(&self) -> f64 {
```
This name is a bit odd because this is not quite a length (it's more of a... percentage? fraction?). But I don't think any of those would be particularly clearer...

Maybe `current_iteration_end_progress` or something?

Also may be worth adding:

```rust
fn current_iteration_duration(&self) -> f64 {
    self.current_iteration_end_progress() * self.duration
}
```

And change the two callers that do that themselves? Just a thought, if you disagree with these suggestions feel free to leave as is, I'm not 100% convinced of them myself :)

mrobinson replied:
I think these suggestions make sense, so I've incorporated them.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#24283.

mrobinson commented:
@bors-servo r=emilio

Thanks again for the reviews!

bors-servo commented:
:pushpin: Commit d319b3b has been approved by `emilio`

<!-- @bors-servo r=emilio d319b3b9ed47031e37f72d2eda03e11282714629 -->

bors-servo commented:
:hourglass: Testing commit d319b3b9ed47031e37f72d2eda03e11282714629 with merge 09c07d0ac9fe99d98644559a957495bf4b8cd83f...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/YHY3PaXaS6Gbw3xoL6N1yA)

jdm commented:
New 2020 passes:
```
  ▶ Unexpected subtest result in /css/css-ui/outline-017.html:
  └ PASS [expected FAIL] outline-color is animated as a color
  ▶ Unexpected subtest result in /css/css-ui/outline-017.html:
  └ PASS [expected FAIL] outline-width is animated as a length
  ▶ Unexpected subtest result in /css/css-ui/outline-017.html:
  └ PASS [expected FAIL] outline-offset is animated as a length
  ▶ Unexpected subtest result in /css/css-ui/outline-018.html:
  └ PASS [expected FAIL] outline-style is animated as a discrete type
```

Failures in 2013 on macOS:
```
  â–¶ Unexpected subtest result in /css/css-animations/animation-iteration-count-009.html:
  â”‚ FAIL [expected PASS] Floating point iteration count during first iteration
  â”‚   â†’ assert_equals: expected "75px" but got "74.9951px"
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/css-animations/animation-iteration-count-009.html:38:16
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  â”‚ promise_test/tests.promise_tests</<@http://web-platform.test:8000/resources/testharness.js:592:36
  â”” promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:591:20

  â–¶ Unexpected subtest result in /css/css-animations/animation-iteration-count-009.html:
  â”‚ FAIL [expected PASS] Floating point iteration count after multiple iterations
  â”‚   â†’ assert_equals: expected "25px" but got "25.0049px"
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/css-animations/animation-iteration-count-009.html:32:16
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  â”‚ promise_test/tests.promise_tests</<@http://web-platform.test:8000/resources/testharness.js:592:36
  â”” promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:591:20

  â–¶ Unexpected subtest result in /css/css-animations/animation-iteration-count-009.html:
  â”‚ FAIL [expected PASS] Floating point iteration count with alternating directions
  â”‚   â†’ assert_equals: expected "25px" but got "25.0049px"
  â”‚ 
  â”‚ @http://web-platform.test:8000/css/css-animations/animation-iteration-count-009.html:44:16
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  â”‚ promise_test/tests.promise_tests</<@http://web-platform.test:8000/resources/testharness.js:592:36
  â”” promise_test/tests.promise_tests<@http://web-platform.test:8000/resources/testharness.js:591:20
```

mrobinson committed:
animations: Don't convert linear easing to a bezier

This conversion can lead to floating point errors and extra work when
computing animations. Avoiding it allows animation-iteration-count-009.html
to pass.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#24283.

mrobinson commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 98a4f55b81a685d6134413af5f41be48cdc88e19 with merge 5b3ede29ad56febd49ac6894075a27a456ecb0ca...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Oty62Q1lRBG_lNHksHpwyA)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#24283.

mrobinson commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit cf4510011fb4bf1ea8f95b7b0874aa952ae99e44 with merge eb56ac004016f45b31c8e18e3aed53fcecd952e7...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/S7gj8P9VR8yGuUhtzgz-Lg)
State: approved= try=True

mrobinson commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit cf4510011fb4bf1ea8f95b7b0874aa952ae99e44 with merge c3a1b471f439caa22aaf54f3e3af98ad93986235...

jdm commented:
The notification got lost, but all tests passed: https://community-tc.services.mozilla.com/tasks/groups/ObJGefmMTfSS3KTT725Ppg

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit cf45100 has been approved by `jdm`

<!-- @bors-servo r=jdm cf4510011fb4bf1ea8f95b7b0874aa952ae99e44 -->

bors-servo commented:
:hourglass: Testing commit cf4510011fb4bf1ea8f95b7b0874aa952ae99e44 with merge 6659e9004dc4941b9aa6164cdc80af61b899e006...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/AZ9LGfurTyKd06qY90ie-A)
Approved by: jdm
Pushing 6659e9004dc4941b9aa6164cdc80af61b899e006 to master...
